Initial Support for ElemRV-N

dnltz opened the pull request:
This pull-request adds initial support for ElemRV-N [0], an end-to-end open-source RISC-V microcontroller designed using SpinalHDL successfully tapped-out with IHP's Open PDK [1].

Support for interfaces already exists on my fork and will be added after this PR got merged.

0: https://github.com/aesc-silicon/elemrv
1: https://github.com/IHP-GmbH/IHP-Open-PDK

dnltz committed:
dts: riscv: Add aesc

ElemRV-N is based on the nitrogen SoC platform. Add the
base nitrogen device-tree and one for elemrv-n. The elemrv-n
device-tree will contain all IP cores later.

Signed-off-by: Daniel Schultz <dnltz@aesc-silicon.de>

dnltz committed:
boards: aesc: Add ElemRV-N

ElemRV-N is an end-to-end open-source microcontroller. This patch
adds basic support for the platform with any interfaces. They will
be added later since drivers are missing completly.

Signed-off-by: Daniel Schultz <dnltz@aesc-silicon.de>

dnltz committed:
soc: Add aesc

Currently, the only available platform is Nitrogen, featuring a
VexRiscv CPU that boots from external SPI flash and runs code from
external HyperRAM.

Signed-off-by: Daniel Schultz <dnltz@aesc-silicon.de>

dnltz committed:
dts: bindings: Add aesc

aesc silicon is a startup focused on developing open-source
silicon solutions.

Link: https://github.com/aesc-silicon/ElemRV

Signed-off-by: Daniel Schultz <dnltz@aesc-silicon.de>

dnltz committed:
MAINTAINERS: Add Aesc Platform

aesc silicon is a startup focused on developing open-source
silicon solutions.

Add a new platform to the MAINTAINERS file to ensure changes
get reviewed.

Signed-off-by: Daniel Schultz <dnltz@aesc-silicon.de>

ycsin reviewed (commented):
the last commit should probably update `doc/releases/release-notes-4.2.rst` as well

henrikbrixandersen commented on boards/others/elemrv/doc/index.rst:
```diff
@@ -0,0 +1,119 @@
+Supported Features
+******************
+
+.. todo:: add board-supported-hw with first interface
```
Why not add it now?

dnltz replied:
Documentation failed to build because of `board-supported-hw directive must be used in a board documentation page.`. I re-added it to take look again.

Maybe @kartben has an idea why it fails?

kartben replied:
You're missing something like this https://github.com/zephyrproject-rtos/zephyr/blob/76e1fc7713a4a3f2b50c497afddec0364a34c10b/boards/others/black_f407ve/doc/index.rst?plain=1#L1

I would recommend using the board documentation template https://github.com/zephyrproject-rtos/zephyr/blob/76e1fc7713a4a3f2b50c497afddec0364a34c10b/doc/templates/board.tmpl

dnltz replied:
Thanks!

dnltz committed:
doc: releases: Add ElemRV to Release Notes

Add a note about initial support for ElemRV-N.

Signed-off-by: Daniel Schultz <dnltz@aesc-silicon.de>

kartben commented on doc/releases/release-notes-4.2.rst:
```diff
@@ -246,6 +246,10 @@ New Boards
 
    * :zephyr:board:`cygnet` (``cygnet``)
 
+* Elements
```
If the vendor of the board is actually "elements" then please put the board in an "elements" folder and not "others". However I couldn't find any reference online to such a vendor?

dnltz replied:
`elements` is not a vendor, it's a project with open source IPs and SoC builder similar to `litex`. `aesc` would be the actual vendor.

henrikbrixandersen replied:
The board should reflect the board vendor. I take it this is the vendors repo? https://github.com/aesc-silicon/ElemRV

dnltz replied:
@henrikbrixandersen yes it is. Updated to `aesc`.

nordicjm commented on soc/elements/vexriscv/Kconfig.soc:
```diff
@@ -0,0 +1,10 @@
+	bool
+
+config SOC_FAMILY
+	default "vexriscv_elements" if SOC_FAMILY_ELEMENTS_VEXRISCV
```
must match e.g. `default "vexriscv_elements" if SOC_FAMILY_VEXRISCV_ELEMENTS` plus must match soc.yml

nordicjm commented on soc/elements/vexriscv/nitrogen/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	default "nitrogen" if SOC_SERIES_NITROGEN
+
+config SOC
+	default "elemrv-n" if SOC_ELEMRV_N
```
underscores not dashes in soc names

dnltz replied:
Is the target name meant to be `board/soc/cpu`? I'm wondering if I should do `elemrv/elemrv/n` instead of `elemrv/elemrv_n`

ycsin replied:
> Is the target name meant to be `board/soc/cpu`?

See [Board terminology](https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#board-terminology)

> I'm wondering if I should do `elemrv/elemrv/n` instead of `elemrv/elemrv_n`

`elemrv/elemrv/n` if the SOC can have different CPU configurations, and most of the other SOC related configurations are the same (i.e. possibility to have 32/64 bit CPUs, or different extensions)

nordicjm commented on soc/elements/vexriscv/soc.yml:
```diff
@@ -0,0 +1,6 @@
+  series:
+  - name: nitrogen
+    socs:
+    - name: elemrv-n
```
underscore not dash

nordicjm commented on dts/riscv/elements/nitrogen.dtsi:
```diff
@@ -0,0 +1,49 @@
+				interrupt-controller;
+			};
+		};
+
```
```suggestion
```

nordicjm commented on boards/others/elemrv/Kconfig.elemrv:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_ELEMRV
+	select SOC_ELEMRV_N if BOARD_ELEMRV_ELEMRV_N
```
```suggestion
	select SOC_ELEMRV_N
```

nordicjm commented on boards/others/elemrv/elemrv_elemrv-n_defconfig:
```diff
@@ -0,0 +1,2 @@
+# Copyright (c) 2025 aesc silicon
```
no uart?

nordicjm replied:
UART should be included in initial submission

dnltz committed:
drivers: serial: Add aesc UART driver

Add minimal support for the aesc silicon UART IP core.

This core includes an internal clock divider and supports flexible
frame configurations, allowing for variable data length, parity, and
stop bit settings.

The current driver version does not support interrupts.

Signed-off-by: Daniel Schultz <dnltz@aesc-silicon.de>

nashif commented on drivers/serial/uart_aesc.c:
```diff
@@ -0,0 +1,125 @@
+{
+	volatile struct uart_aesc_regs *uart = DEV_UART(dev);
+
+	while ((uart->status & 0x00FF0000) == 0) {
```
Assign this magic number to a well-named variable, and use the variable instead.

nashif commented on drivers/serial/uart_aesc.c:
```diff
@@ -0,0 +1,125 @@
+	int val;
+
+	val = uart->read_write;
+	if (val & 0x10000) {
```
Assign this magic number to a well-named variable, and use the variable instead.

dnltz commented:
@nordicjm @ycsin @kartben sorry for staling your reviews. I added an uart driver to this pull-request to build and run the `hello_world` sample.

henrikbrixandersen commented on drivers/serial/uart_aesc.c:
```diff
@@ -0,0 +1,129 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(aesc_uart, CONFIG_UART_LOG_LEVEL);
+
+#define DEV_UART_IRQ_TX_EN		(1 << 0)
+#define DEV_UART_IRQ_RX_EN		(1 << 1)
```
Please use the `BIT()` macro.

henrikbrixandersen commented on drivers/serial/uart_aesc.c:
```diff
@@ -0,0 +1,129 @@
+#define DEV_UART_IRQ_RX_EN		(1 << 1)
+
+struct uart_aesc_data {
+	DEVICE_MMIO_NAMED_RAM(regs);
```
Will this driver be present on SoC with an MMU? Otherwise, this is not needed.

dnltz replied:
Not yet but eventually in the future. So, if it's not a problem I would keep it.

henrikbrixandersen commented on drivers/serial/uart_aesc.c:
```diff
@@ -0,0 +1,129 @@
+#define DEV_UART(dev)							     \
+	((struct uart_aesc_regs *)DEVICE_MMIO_NAMED_GET(dev, regs))
+
+#define		AESC_UART_FIFO_TX_COUNT_MASK		0x00FF0000
```
Please use `GENMASK()`.

henrikbrixandersen commented on drivers/serial/uart_aesc.c:
```diff
@@ -0,0 +1,129 @@
+	((struct uart_aesc_regs *)DEVICE_MMIO_NAMED_GET(dev, regs))
+
+#define		AESC_UART_FIFO_TX_COUNT_MASK		0x00FF0000
+#define		AESC_UART_READ_FIDO_VALID_BIT		0x10000
```
Please use `BIT()`.

henrikbrixandersen commented on drivers/serial/uart_aesc.c:
```diff
@@ -0,0 +1,129 @@
+	volatile struct uart_aesc_regs *uart;
+
+	DEVICE_MMIO_NAMED_MAP(dev, regs, K_MEM_CACHE_NONE);
+	LOG_INF("IP core version: %i.%i.%i.",
```
Reduce the log level to `LOG_DBG()`.

henrikbrixandersen commented on drivers/serial/uart_aesc.c:
```diff
@@ -0,0 +1,129 @@
+		ip_id_get_patchlevel(base_addr)
+	);
+	DEVICE_MMIO_NAMED_GET(dev, regs) = ip_id_relocate_driver(base_addr);
+	LOG_INF("Relocate driver to address 0x%lx.", DEVICE_MMIO_NAMED_GET(dev, regs));
```
Same here.

henrikbrixandersen commented on drivers/serial/uart_aesc.c:
```diff
@@ -0,0 +1,129 @@
+
+	return 0;
+}
+
```
```suggestion
```

henrikbrixandersen commented on dts/bindings/serial/aesc,uart.yaml:
```diff
@@ -0,0 +1,18 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: aesc UART - This binding gives a base representation of the aesc UART
```
Please use the newly introduced `title` key as well as `description` and simplify the text accordingly.

henrikbrixandersen commented on dts/bindings/serial/aesc,uart.yaml:
```diff
@@ -0,0 +1,18 @@
+
+compatible: "aesc,uart"
+
+# inherits:
```
```suggestion
```

henrikbrixandersen commented on dts/bindings/serial/aesc,uart.yaml:
```diff
@@ -0,0 +1,18 @@
+
+description: aesc UART - This binding gives a base representation of the aesc UART
+
+compatible: "aesc,uart"
```
Is this peripheral a general peripheral from `aesc` or is it specific to the ElemRV-N?

dnltz replied:
General peripheral but used in ElemRV.

henrikbrixandersen commented on dts/bindings/serial/aesc,uart.yaml:
```diff
@@ -0,0 +1,18 @@
+
+properties:
+  reg:
+    type: array
+    description: mmio register space
```
Do not override `base.yaml` properties.

```suggestion
```

henrikbrixandersen commented on dts/bindings/vendor-prefixes.txt:
```diff
@@ -33,6 +33,7 @@ adh	AD Holdings Plc.
 adi	Analog Devices, Inc.
 advantech	Advantech Corporation
 aeroflexgaisler	Aeroflex Gaisler AB
+aesc	aesc silicon
```
All lower-case?

dnltz replied:
Changed the name everywhere and capitalized it.

henrikbrixandersen commented on dts/riscv/aesc/nitrogen.dtsi:
Here it's called `nitrogen` elsewhere it's simply `n`? Can we use `nitrogen` all-over?

dnltz replied:
I would like to stick to `ElemRV-N` at board level since that's the actual chip name. I'm fine with renaming the device-tree to `elemrn-nitrogen.dtsi` if that makes it better.

henrikbrixandersen commented on include/zephyr/drivers/aesc/ip_identification.h:
Quite a few magic numbers involved here. I'd suggest converting to using `#define` and `FIELD_GET()`.

henrikbrixandersen commented on soc/aesc/vexriscv/nitrogen/Kconfig.defconfig:
```diff
@@ -0,0 +1,18 @@
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default $(dt_node_int_prop_int,/cpus/cpu@0,clock-frequency)
+
+config SYS_CLOCK_TICKS_PER_SEC
+	default 1000
```
Why change this?

henrikbrixandersen commented on soc/aesc/vexriscv/Kconfig.soc:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2025 aesc silicon
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_FAMILY_AESC_VEXRISCV
```
I'm a bit confused by the whole `aesc`, `vexrisc`, `nitrogen`, `elemrv`, `n` naming here. Is the family really vexriscv?

dnltz replied:
Actually after checking this comment I figured out that `SOC_FAMILY_AESC_VEXRISCV` is wrong and not required at all. I removed all these configs here.

dnltz commented on boards/aesc/elemrv/elemrv_elemrv_n.dts:
```diff
@@ -0,0 +1,50 @@
+};
+
+&uart0 {
+	clock-frequency = <DT_FREQ_M(20)>;
```
Yes, that chip has currently only one clock-domain, but will have multiple in the future.

Only UART needs the clock-frequency to calculate the baud rate. So, I'm not too afraid it doesn't scale and in future versions a clock-controller should provide this information.

ycsin commented on include/zephyr/drivers/aesc/ip_identification.h:
```diff
@@ -0,0 +1,100 @@
+/** @file
```
This seems too vendor-specific to be placed here, should this be in `soc/aesc/...` and exposed via `zephyr_include_directories(.)` instead?

dnltz replied:
good idea! I was not very happy about the location either.

fkokosinski commented on dts/riscv/aesc/elemrv-n.dtsi:
```diff
@@ -0,0 +1,26 @@
+			reg = <0xf0006000 0x1000>;
+			status = "disabled";
+		};
+
```
Nit: unnecessary blank line

fkokosinski commented on soc/aesc/vexriscv/nitrogen/Kconfig:
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2025 Aesc Silicon
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_NITROGEN
```
The path of this (and other) files should probably be `soc/aesc/nitrogen/Kconfig`

fkokosinski commented on soc/aesc/vexriscv/soc.yml:
```diff
@@ -0,0 +1,6 @@
+family:
+- name: aesc_vexriscv
```
`aesc_vexriscv` doesn't sound like SoC family. It's just the vendor name + CPU name. Please note that the `family` property is not required.

dnltz replied:
I didn't know that. Thanks! Removed the `family` property and replaced `vexriscv` with the  `nitrogen` directory.
